// @arco-design/web-vue/lib/_components/virtual-list-v2/virtual-list-item.js
"use strict";
var vue = require("vue");
var vueUtils = require("../../_utils/vue-utils.js");
var VirtualListItem = vue.defineComponent({
  name: "VirtualListItem",
  props: {
    hasItemSize: {
      type: Function,
      required: true
    },
    setItemSize: {
      type: Function,
      required: true
    }
  },
  setup(props, {
    slots
  }) {
    var _a;
    const key = (_a = vue.getCurrentInstance()) == null ? void 0 : _a.vnode.key;
    const itemRef = vue.ref();
    const setItemSize = () => {
      var _a2, _b;
      const ele = (_b = (_a2 = itemRef.value) == null ? void 0 : _a2.$el) != null ? _b : itemRef.value;
      if (!props.hasItemSize(key) && (ele == null ? void 0 : ele.offsetHeight)) {
        props.setItemSize(key, ele.offsetHeight);
      }
    };
    vue.onMounted(() => setItemSize());
    return () => {
      var _a2;
      const child = vueUtils.getFirstComponent((_a2 = slots.default) == null ? void 0 : _a2.call(slots));
      if (child) {
        return vue.cloneVNode(child, {
          ref: itemRef
        }, true);
      }
      return null;
    };
  }
});
module.exports = VirtualListItem;
