// tdesign-vue-next/es/_common/js/tree/tree-node-model.mjs
/**
 * tdesign v0.24.2
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../../../_chunks/dep-3cecb2d2.mjs';
import { p as pick_1 } from '../../../_chunks/dep-e3e7f0af.mjs';
import { o as omit_1 } from '../../../_chunks/dep-2289c323.mjs';
import log from '../log/log.mjs';
import '../../../_chunks/dep-ad7eea53.mjs';
import '../../../_chunks/dep-86de30fa.mjs';
import '../../../_chunks/dep-de110986.mjs';
import '../../../_chunks/dep-21468dc9.mjs';
import '../../../_chunks/dep-eede90e8.mjs';
import '../../../_chunks/dep-f5995493.mjs';
import '../../../_chunks/dep-63f67da3.mjs';
import '../../../_chunks/dep-cab779bd.mjs';
import '../../../_chunks/dep-8c2e57f5.mjs';
import '../../../_chunks/dep-c5944b2d.mjs';
import '../../../_chunks/dep-a39da6ce.mjs';
import '../../../_chunks/dep-f5b976b2.mjs';
import '../../../_chunks/dep-d904f554.mjs';
import '../../../_chunks/dep-22081ace.mjs';
import '../../../_chunks/dep-0fdea1fe.mjs';
import '../../../_chunks/dep-a6ebc045.mjs';
import '../../../_chunks/dep-6f8d8885.mjs';
import '../../../_chunks/dep-2e0d3d32.mjs';
import '../../../_chunks/dep-6ad4a916.mjs';
import '../../../_chunks/dep-cd44ef02.mjs';
import '../../../_chunks/dep-a7f196e9.mjs';
import '../../../_chunks/dep-5a93c66b.mjs';
import '../../../_chunks/dep-c692ab23.mjs';
import '../../../_chunks/dep-a8762670.mjs';
import '../../../_chunks/dep-b43427a8.mjs';
import '../../../_chunks/dep-1984c62f.mjs';
import '../../../_chunks/dep-a537ee3e.mjs';
import '../../../_chunks/dep-e478235e.mjs';
import '../../../_chunks/dep-dcaa2bc6.mjs';
import '../../../_chunks/dep-0510ab65.mjs';
import '../../../_chunks/dep-e23bac6d.mjs';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function getExposedProps(node) {
  var props = pick_1(node, ["value", "label", "data", "actived", "expanded", "checked", "indeterminate", "loading"]);
  return props;
}
function createNodeModel(node) {
  var props = getExposedProps(node);
  var model = _objectSpread(_objectSpread({}, props), {}, {
    getLevel: function getLevel() {
      return node.getLevel();
    },
    getIndex: function getIndex() {
      return node.getIndex();
    },
    isFirst: function isFirst() {
      return node.isFirst();
    },
    isLast: function isLast() {
      return node.isLast();
    },
    isLeaf: function isLeaf() {
      return node.isLeaf();
    },
    insertBefore: function insertBefore(newData) {
      return node.insertBefore(newData);
    },
    insertAfter: function insertAfter(newData) {
      return node.insertAfter(newData);
    },
    appendData: function appendData(data) {
      return node.append(data);
    },
    getPath: function getPath() {
      var nodes = node.getPath();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getParent: function getParent() {
      var _node$parent;
      return (_node$parent = node.parent) === null || _node$parent === void 0 ? void 0 : _node$parent.getModel();
    },
    getParents: function getParents() {
      var nodes = node.getParents();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getRoot: function getRoot() {
      var root = node.getRoot();
      return root === null || root === void 0 ? void 0 : root.getModel();
    },
    getSiblings: function getSiblings() {
      var nodes = node.getSiblings();
      return nodes.map(function (item) {
        return item.getModel();
      });
    },
    getChildren: function getChildren(deep) {
      var childrenModel = false;
      var children = node.children;
      if (Array.isArray(children)) {
        if (children.length > 0) {
          if (deep) {
            var nodes = node.walk();
            nodes.shift();
            childrenModel = nodes.map(function (item) {
              return item.getModel();
            });
          } else {
            childrenModel = children.map(function (item) {
              return item.getModel();
            });
          }
        } else {
          childrenModel = false;
        }
      } else if (typeof children === "boolean") {
        childrenModel = children;
      }
      return childrenModel;
    },
    remove: function remove(value) {
      if (!value) {
        node.remove();
        return;
      }
      var tree = node.tree;
      var targetNode = tree.getNode(value);
      if (!targetNode) {
        log.warnOnce("Tree", "`".concat(value, "` is not exist"));
        return;
      }
      var parents = targetNode.getParents();
      var parentValues = parents.map(function (pnode) {
        return pnode.value;
      });
      if (parentValues.indexOf(node.value) < 0) {
        log.warnOnce("Tree", "`".concat(value, "` is not a childNode of current node"));
        return;
      }
      targetNode.remove();
    },
    setData: function setData(data) {
      var _data = omit_1(data, ["children", "value", "label"]);
      var keys = node.tree.config.keys;
      var dataValue = data[(keys === null || keys === void 0 ? void 0 : keys.value) || "value"];
      var dataLabel = data[(keys === null || keys === void 0 ? void 0 : keys.label) || "label"];
      if (dataValue !== void 0) _data.value = dataValue;
      if (dataLabel !== void 0) _data.label = dataLabel;
      Object.assign(node.data, _data);
      Object.assign(node, _data);
    }
  });
  return model;
}
function updateNodeModel(model, node) {
  var props = getExposedProps(node);
  Object.assign(model, props);
}

export { createNodeModel, updateNodeModel };
