// tdesign-vue-next/es/tree/constants.mjs
/**
 * tdesign v0.24.2
 * (c) 2022 tdesign
 * @license MIT
 */

import { computed } from 'vue';
import { usePrefixClass } from '../hooks/useConfig.mjs';
import '../config-provider/useConfig.mjs';
import '../_chunks/dep-a953bb61.mjs';
import '../_chunks/dep-ad7eea53.mjs';
import '../_chunks/dep-b43427a8.mjs';
import '../_chunks/dep-cab779bd.mjs';
import '../_chunks/dep-8c2e57f5.mjs';
import '../_chunks/dep-c5944b2d.mjs';
import '../_chunks/dep-f5995493.mjs';
import '../_chunks/dep-21468dc9.mjs';
import '../_chunks/dep-a39da6ce.mjs';
import '../_chunks/dep-f5b976b2.mjs';
import '../_chunks/dep-d904f554.mjs';
import '../_chunks/dep-1984c62f.mjs';
import '../_chunks/dep-cd44ef02.mjs';
import '../_chunks/dep-63f67da3.mjs';
import '../_chunks/dep-a537ee3e.mjs';
import '../_chunks/dep-a6ebc045.mjs';
import '../_chunks/dep-6f8d8885.mjs';
import '../_chunks/dep-6ad4a916.mjs';
import '../_chunks/dep-de110986.mjs';
import '../_chunks/dep-2e0d3d32.mjs';
import '../_chunks/dep-e478235e.mjs';
import '../_chunks/dep-a8f3adf1.mjs';
import '../_chunks/dep-c692ab23.mjs';
import '../_chunks/dep-e23bac6d.mjs';
import '../_chunks/dep-6d55d809.mjs';
import '../_common/js/global-config/default-config.mjs';
import '../_common/js/global-config/locale/zh_CN.mjs';
import '../config-provider/type.mjs';

var useCLASSNAMES = function useCLASSNAMES() {
  var classPrefix = usePrefixClass();
  return computed(function () {
    var tree = "".concat(classPrefix.value, "-tree");
    return {
      icon: "".concat(classPrefix.value, "-icon"),
      folderIcon: "".concat(classPrefix.value, "-folder-icon"),
      actived: "".concat(classPrefix.value, "-is-active"),
      disabled: "".concat(classPrefix.value, "-is-disabled"),
      treeIconRight: "".concat(classPrefix.value, "-icon-arrow-right"),
      treeIconDown: "".concat(classPrefix.value, "-icon-arrow-down"),
      treeNodeToggle: "".concat(classPrefix.value, "-tree-toggle"),
      tree: tree,
      treeTransition: "".concat(tree, "--transition"),
      treeBlockNode: "".concat(tree, "--block-node"),
      treeEmpty: "".concat(tree, "__empty"),
      treeList: "".concat(tree, "__list"),
      treeNode: "".concat(tree, "__item"),
      treeNodeOpen: "".concat(tree, "__item--open"),
      treeHoverAble: "".concat(tree, "--hoverable"),
      treeCheckable: "".concat(tree, "--checkable"),
      treeLabel: "".concat(tree, "__label"),
      treeLabelStrictly: "".concat(tree, "__label--strictly"),
      treeIcon: "".concat(tree, "__icon"),
      treeIconDefault: "".concat(tree, "__icon--default"),
      treeSpace: "".concat(tree, "__space"),
      treeOperations: "".concat(tree, "__operations"),
      treeNodeVisible: "".concat(tree, "__item--visible"),
      treeNodeHidden: "".concat(tree, "__item--hidden"),
      treeNodeEnter: "".concat(tree, "__item--enter-active"),
      treeNodeLeave: "".concat(tree, "__item--leave-active"),
      treeNodeDraggable: "".concat(tree, "__item--draggable"),
      treeNodeDragging: "".concat(tree, "__item--dragging"),
      treeNodeDragTipTop: "".concat(tree, "__item--tip-top"),
      treeNodeDragTipBottom: "".concat(tree, "__item--tip-bottom"),
      treeNodeDragTipHighlight: "".concat(tree, "__item--tip-highlight"),
      line: "".concat(tree, "__line"),
      lineIsLeaf: "".concat(tree, "__line--leaf"),
      lineIsFirst: "".concat(tree, "__line--first")
    };
  });
};
var injectKey = Symbol("treeProvider");
var dragInjectKey = Symbol("onDrag");

export { dragInjectKey, injectKey, useCLASSNAMES };
