// tdesign-vue-next/esm/time-picker/props.js
/**
 * tdesign v0.24.2
 * (c) 2022 tdesign
 * @license MIT
 */

var props = {
  allowInput: Boolean,
  clearable: Boolean,
  disabled: Boolean,
  disableTime: {
    type: Function
  },
  format: {
    type: String,
    "default": "HH:mm:ss"
  },
  hideDisabledTime: {
    type: Boolean,
    "default": true
  },
  inputProps: {
    type: Object
  },
  placeholder: {
    type: String,
    "default": void 0
  },
  popupProps: {
    type: Object
  },
  size: {
    type: String,
    "default": "medium",
    validator: function validator(val) {
      if (!val) return true;
      return ["small", "medium", "large"].includes(val);
    }
  },
  steps: {
    type: Array,
    "default": function _default() {
      return [1, 1, 1];
    }
  },
  value: {
    type: String,
    "default": void 0
  },
  modelValue: {
    type: String,
    "default": void 0
  },
  defaultValue: {
    type: String,
    "default": ""
  },
  onBlur: Function,
  onChange: Function,
  onClose: Function,
  onFocus: Function,
  onInput: Function,
  onOpen: Function
};

export { props as default };
